### ISSUE TITLE
Delimited Text Ingest Fails on Unescaped Quotes

### ISSUE THREAD

**Issue by ezbc** [marklogic/marklogic-contentpump#68]
# Issue 
When importing delimited text MLCP attempts to skip ill-formatted records with unescaped quotes. 

MLCP skips a poorly-formatted three-column row like the following:
```"cell 2 1","cell "2" 2","cell 2 3"```

but fails entirely on a poorly-formatted three-column row like the following:
```"cell 3 1","cell "3", "2",","cell 3 3"```

with the error: `ERROR contentpump.ContentPump: Error running a ContentPump job
java.lang.RuntimeException: java.io.IOException: (line 5) invalid char between encapsulated token and delimiter`

When ingesting the following file I would expect records 1, 4, and 5 to be ingested, and records 2 and 3 to be skipped:
```
column1,column2,column3
"cell 1 1","cell 1 2","cell 1 3"
"cell 2 1","cell "2" 2","cell 2 3"
"cell 3 1","cell "3", "2",","cell 3 3"
"cell 4 1","cell 4 2","cell 4 3"
"cell 5 1","cell 5 2","cell 5 3"
```
however mlcp throws a fatal error when ingesting record 3. Here is a list of the documents ingested:
![image](https://user-images.githubusercontent.com/6400323/33279708-2ee71544-d365-11e7-8b1b-bbf1747c596c.png). Only the first record was ingested. The second record was skipped. The job failed on the third record. Records 4 and 5 were not ingested.

## Materials
A stack trace, an example csv, and a config file are provided in the attached content 
[issue-report.zip](https://github.com/marklogic/marklogic-contentpump/files/1506891/issue-report.zip). A thread count of 1 is intentional in order to show the stack trace in stdout.

## Environment Details
MarkLogic Server Version: 9.0-2
ContentPump version: 9.0.2
Java version: 1.8.0_152

**Comment by ezbc** [comment]
@jxchen-us thanks for responding to this. I believe the error arises because the [DelimitedTextReader](https://github.com/marklogic/marklogic-contentpump/blob/d5239d39a90a2da8b220a7ed764cf2e0dd6914d3/mlcp/src/main/java/com/marklogic/contentpump/DelimitedTextReader.java#L274)
 is not prepared to catch an error when it looks for the next column in the row.

**Comment by mattsunsjf** [comment]
Thumbs up for very detailed and organized issue report!

**Comment by mattsunsjf** [comment]
It's good that you could see how this CSV is poorly formatted. There are probably tens of ways that a CSV can be poorly formatted. Your sample file may have catch a couple of them.

Mlcp tries to handle as graceful as possible the poor format, so the skipped the case falls into this category. The failure lines, the reason for errors are complicated. But the basic idea is, mlcp uses Apache's CSVParser to parse files. Mlcp does not parse and skip the CSV record directly, because the library does a better job. There are some known limitations of CSV libraries that they are able to handle broken CSV files. I'm not only referring to the CSV library we are using, but also others. We have tried a few. There is not a good parse as I know can handle ALL kinds of broken CSV, as I said, there are probably tens of ways that a CSV can be poorly formatted.

Although above said, there still might be ways to achieve what you expected. First, can you grab the latest mlcp version 9.0-3? There is a critical fix went in 9.0-3 that deals with CSVParser. Secondly, try split input. So that lines will be handled separately. You may want to fine tune the split size to make each line a split.

Another option for you is to correct the CSV lines. You can simply drop the broken line, or fix the broken fields.
